extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3StorageDruidModule.java

```diff
@@ -182,4 +186,16 @@ public Supplier<ServerSideEncryptingAmazonS3> getAmazonS3ClientSupplier(
   {
     return Suppliers.memoize(serverSideEncryptingAmazonS3Builder::build);
   }
+
+  @Provides
+  @LazySingleton
+  public S3UploadManager getS3UploadManager(ScheduledExecutorFactory scheduledExecutorFactory, RuntimeInfo runtimeInfo)
+  {
+    int poolSize = Math.max(4, runtimeInfo.getAvailableProcessors());
+    ScheduledExecutorService executorService = scheduledExecutorFactory.create(
+        poolSize,
+        "UploadThreadPool-%d"
+    );
+    return new S3UploadManager(executorService);
+  }
```

Comment:
Instead of this, I guess it would be better to just have an `@Inject` constructor for `S3UploadManager`. That way, the executor service would be created inside the `S3UploadManager`, giving it complete control over its lifecycle.